builtin/v13/migration/market.go

```diff
@@ -0,0 +1,619 @@
+	// providerSectorsRoot: HAMT[ActorID]HAMT[SectorNumber]SectorDealIDs
+
+	okIn, prevInStates, err := in.Cache.Read(migration.MarketPrevDealStatesInKey(in.Address))
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get previous inRoot from cache: %w", err)
+	}
+
+	okInPr, prevInProposals, err := in.Cache.Read(migration.MarketPrevDealProposalsInKey(in.Address))
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get previous inRoot from cache: %w", err)
+	}
+
+	okOut, prevOutStates, err := in.Cache.Read(migration.MarketPrevDealStatesOutKey(in.Address))
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get previous outRoot from cache: %w", err)
+	}
+
+	okOutPs, prevOutProviderSectors, err := in.Cache.Read(migration.MarketPrevProviderSectorsOutKey(in.Address))
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get previous outRoot from cache: %w", err)
+	}
+
+	var providerSectorsRoot, newStateArrayRoot cid.Cid
+
+	if okIn && okInPr && okOut && okOutPs {
+		providerSectorsRoot, newStateArrayRoot, err = m.migrateProviderSectorsAndStatesWithDiff(ctx, store, prevInStates, prevOutStates, prevOutProviderSectors, states, prevInProposals, proposals)
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to migrate provider sectors (diff): %w", err)
+		}
+	} else {
+		providerSectorsRoot, newStateArrayRoot, err = m.migrateProviderSectorsAndStatesFromScratch(ctx, store, in, states)
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to migrate provider sectors (all): %w", err)
+		}
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevDealStatesInKey(in.Address), states); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous inRoot to cache: %w", err)
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevDealProposalsInKey(in.Address), proposals); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous inRoot to cache: %w", err)
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevDealStatesOutKey(in.Address), newStateArrayRoot); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous outRoot to cache: %w", err)
+	}
+
+	if err := in.Cache.Write(migration.MarketPrevProviderSectorsOutKey(in.Address), providerSectorsRoot); err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to write previous outRoot to cache: %w", err)
+	}
+
+	return providerSectorsRoot, newStateArrayRoot, nil
+}
+
+func (m *marketMigrator) migrateProviderSectorsAndStatesWithDiff(ctx context.Context, store cbor.IpldStore, prevInStatesCid, prevOutStatesCid, prevOutProviderSectorsCid, inStatesCid, prevInProposals, inProposals cid.Cid) (cid.Cid, cid.Cid, error) {
+	diffs, err := amt.Diff(ctx, store, store, prevInStatesCid, inStatesCid, amt.UseTreeBitWidth(market12.StatesAmtBitwidth))
```

Comment:
So, unfortunately, we're diffing the wrong states here if we care about re-orgs. I.e., it's possible for a re-org to change the sector number to which a deal is assigned without changing anything about the deal itself, and this diff won't catch that because it's diffing the deal states which don't record sector information.

I discussed this with @arajasek and we think the answer is to just not allow reorgs. I.e., don't use the cached migration if its not an ancestor. That'll ensure that neither sector nor deal IDs are reused.

Reply:
Basically:

1. Wait for some measure of finality (not real finality, but, like, a few epochs) after the pre-migration epoch before running the pre-migration. OR, run the premigration at the specified epoch, but do a lookback (i.e., migrate state from a few epochs ago).
2. In the "migration" struct itself, record the highest pre-migration tipset.
3. When running the actual migration (or, really, any pre-migrations), if there is a pre-migration tipset recorded and it's not an ancestor of the tipset being migrated, discard the cache.

Reply:
Ugh. Ok, maybe that's not the way. We _do_ diff the sectors anyways. So...

1. When we diff the sectors, we need to record the changes.
2. When we diff the deal states, we need to apply/remove the sector changes.
3. Finally, we need to iterate over the _rest_ of the sector changes, and fix their deal states.

Reply:
@magik6k is the above note something you can look at addressing?